tests/handlers/test_appservice.py

```diff
@@ -412,6 +413,146 @@ def prepare(self, reactor, clock, hs):
+        # This is normally set in `notify_interested_services` but we need to call the
+        # internal async version so the reactor gets pushed to completion.
+        self.hs.get_application_service_handler().current_max += 1
+        self.get_success(
+            self.hs.get_application_service_handler()._notify_interested_services(
+                RoomStreamToken(
+                    None, self.hs.get_application_service_handler().current_max
+                )
+            )
+        )
+
+    def test_match_local_room_members(self):
+        """
+        Test to make sure that a user local to the server and in the room is notified
+        when someone else in the room sends a message.
+        """
+        # Register an application service that's interested in local and remote user
+        interested_appservice = self._register_application_service(
+            namespaces={
+                ApplicationService.NS_USERS: [
+                    {
+                        "regex": "@local_as_user:test",
+                        "exclusive": True,
+                    },
+                ],
+            },
+        )
+
+        alice = self.register_user("alice", "pass")
+        alice_access_token = self.login("alice", "pass")
+        room_id = self.helper.create_room_as(room_creator=alice, tok=alice_access_token)
+
+        # Join the interesting user to the room
+        self.get_success(
+            event_injection.inject_member_event(
+                self.hs, room_id, "@local_as_user:test", "join"
+            )
+        )
+        # Kick the appservice into checking this membership event to get it out of the
+        # way
+        self._notify_interested_services()
+        # We don't care about the interesting user join event (this test is making sure
+        # the next thing works)
+        self.send_mock.reset_mock()
+
+        # Send a message from an uninteresting user
+        self.helper.send_event(
+            room_id,
+            type=EventTypes.Message,
+            content={
+                "msgtype": "m.text",
+                "body": "message from uninteresting user",
+            },
+            tok=alice_access_token,
+        )
+        # Kick the appservice into checking this new event
+        self._notify_interested_services()
+
+        self.send_mock.assert_called_once()
+        (
+            service,
+            events,
+            _ephemeral,
+            _to_device_messages,
+            _otks,
+            _fbks,
+            _device_list_summary,
+        ) = self.send_mock.call_args[0]
+
+        # Even though the message came from an uninsteresting user, it should still
+        # notify us because the interesting user is joined to the room.
+        self.assertEqual(service, interested_appservice)
+        self.assertEqual(events[0]["type"], "m.room.message")
+        self.assertEqual(events[0]["sender"], alice)
+
+    def test_match_remote_room_members(self):
```

Comment:
These tests are practically the same. Only difference is the user namespace regex and the interesting user itself.

We could change this to a for-loop:

```py
for interesting_user in ["@local_as_user:test", "@interesting_user:remote"]:
    def test_match_interesting_room_members(self):
       ...
```

Reply:
Done using `@parameterized.expand(["@local_as_user:test", "@interesting_user:remote"])`